osbuild/buildroot.py

```diff
@@ -263,3 +267,23 @@ def run(self, argv, monitor, binds=None, readonly_binds=None):
         data.close()
 
         return CompletedBuild(proc, output)
+
+    @classmethod
+    def read_with_timeout(cls, proc, poller, start, timeout=600000):
+        remaining = timeout - (time.monotonic() - start)
+        if remaining <= 0:
+            proc.terminate()
+            raise TimeoutError
+
+        buf = None
+        fd = proc.stdout.fileno()
+        events = poller.poll(remaining)
+        for fd, flag in events:
+            if flag & (select.POLLIN | select.POLLPRI):
+                buf = os.read(fd, 32768)
+            elif flag & (select.POLLHUP | select.POLLERR):
+                proc.terminate()
```

Comment:
This changes the behaviour a bit from before (we would not explicitly terminate when the socket was closed). How come we didn't need this in the past? Or was that a bug?

Reply:
Yeah your right, 
My thinking here was that as 3rd party rpm's get added then I should be more harsh on any errors.

I can get rid of this `elif` altogether if its not what ye guys want :smile:

Reply:
I gave this a bit more thought, just a couple of things to think about / verify:
- this handles `POLLHUP` in an `elif` from `POLLIN`, so if you get both you end up not handling `POLLHUP`, which might be fine depending on if this stuff is edge or level triggered, but i hate it when i have to think...
- it feels a bit ugly to send a `SIGTERM` on `POLLHUP`.... what if everything is cleaning up and suddenly the stage gets terminated when it was doing an orderly shutdown, probably will not work out like that, but still felt iffy
- we do have to do a `SIGTERM` at some point in the case a `POLLHUP` is sent, in case the stage keeps working after that, but really we should keep the timeout running and only do it when it is hit, independently of what happens to the socket

Reply:
I see what you mean,

It seems like the best option would be to drop the `POLLHUP` check altogether. As you said, the timeout will catch a hanging socket regardless. 

If your happy with the above I can remove `POLLHUP` and keep everything else the same

Reply:
Hm, but then we would always hang until the timeout is out, even if the stage exited cleanly, wouldn't we?

Reply:
Yep, I just saw that,
Moving `POLLHUP` to the first `if` statement produces the correct behaviour though, and seems to be a good solution.

Reply:
I think that still has a problem as it would allow someone to close the socket (giving us a `POLLHUP`), but never actually exiting the service.

To get this unblocked, how about we keep it as it is now, but make the `elif` into an `if` for the `POLLHUP` handling, so we make sure we don't miss it. Then we can do a follow-up, if necessary along the lines @gicmo proposed in his comment below.

Reply:
Done :smile: